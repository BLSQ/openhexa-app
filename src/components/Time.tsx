import { DateTime, DateTimeOptions } from "luxon";
import { memo } from "react";

type Props = {
  datetime: string;
  className?: string;
  format?: DateTimeOptions;
};

const Time = (props: Props) => {
  const datetime = DateTime.fromISO(props.datetime);

  return (
    <time
      title={datetime.toISO()}
      dateTime={datetime.toISO()}
      className={props.className}
    >
      {datetime.toLocaleString(props.format ?? DateTime.DATETIME_SHORT)}
    </time>
  );
};

export default memo(Time);
