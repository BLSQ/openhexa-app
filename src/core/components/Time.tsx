import clsx from "clsx";
import useInterval from "core/hooks/useInterval";
import useRelativeTime from "core/hooks/useRelativeTime";
import { DateTime, DateTimeOptions } from "luxon";
import { memo, useEffect, useMemo, useState } from "react";

type Props = {
  datetime: string;
  className?: string;
  relative?: boolean;
  format?: DateTimeOptions;
};

const Time = (props: Props) => {
  const datetime = useMemo(
    () => DateTime.fromISO(props.datetime),
    [props.datetime]
  );

  const relativeDate = useRelativeTime(datetime);

  const value = useMemo(() => {
    if (props.relative) {
      return relativeDate;
    } else {
      return datetime.toLocaleString(props.format ?? DateTime.DATETIME_SHORT);
    }
    // We use the toggle variable to force React to recompute this value
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [datetime, props.format, props.relative, relativeDate]);

  return (
    <time
      /* Relative time can quickly change between the server rendering and the client rendering. Let's ignore this while hydrating the dom */
      suppressHydrationWarning={props.relative}
      title={datetime.toISO()}
      dateTime={datetime.toISO()}
      className={clsx("whitespace-nowrap", props.className)}
    >
      {value}
    </time>
  );
};

export default memo(Time);
