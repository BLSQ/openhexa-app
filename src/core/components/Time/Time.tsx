import clsx from "clsx";
import useRelativeTime from "core/hooks/useRelativeTime";
import { DateTime, DateTimeOptions } from "luxon";
import { useMemo } from "react";

type Props = {
  datetime: string;
  className?: string;
  relative?: boolean;
  format?: DateTimeOptions;
};

const Time = (props: Props) => {
  const datetime = useMemo(
    // By default, all dates from the backend are in UTC
    () => DateTime.fromISO(props.datetime).toLocal(),
    [props.datetime],
  );

  const relativeDate = useRelativeTime(datetime);

  if (!datetime?.isValid) return null;

  const isoDate = datetime.toISO();
  return (
    <time
      suppressHydrationWarning={true}
      title={isoDate ?? undefined}
      dateTime={isoDate ?? undefined}
      className={clsx("whitespace-nowrap", props.className)}
    >
      {props.relative
        ? relativeDate
        : datetime.toLocaleString(props.format ?? DateTime.DATETIME_SHORT)}
    </time>
  );
};

export default Time;
