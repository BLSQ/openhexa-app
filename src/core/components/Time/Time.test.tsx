import Time from "./Time";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DateTime, Settings } from "luxon";

const nowMock = Settings.now as jest.Mock;

describe("Time", () => {
  it("renders the time in different formats", async () => {
    const dt = DateTime.local(2022, 2, 20, 10, 11, 34);
    Settings.defaultLocale = "en";
    const { container, rerender } = render(<Time datetime={dt.toISO()} />);
    expect(container).toMatchSnapshot();

    rerender(<Time datetime={dt.toISO()} format={DateTime.DATE_SHORT} />);
    expect(container).toMatchSnapshot();
  });

  it("renders relative time", async () => {
    const dt = DateTime.local(2022, 2, 20, 10, 11, 34);
    nowMock.mockReturnValue(dt.plus({ hours: 0, minutes: 10 }).toMillis());
    const { container, rerender } = render(
      <Time relative datetime={dt.toISO()} />
    );
    expect(container).toMatchSnapshot();

    nowMock.mockReturnValue(dt.plus({ hours: 2, minutes: 30 }).toMillis());
    rerender(<Time relative datetime={dt.toISO()} />);
    expect(container).toMatchSnapshot();
  });
});
