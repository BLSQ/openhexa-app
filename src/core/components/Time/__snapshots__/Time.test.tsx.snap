// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time renders relative time 1`] = `
<div>
  <time
    class="whitespace-nowrap"
    datetime="2022-02-20T10:11:34.000+01:00"
    title="2022-02-20T10:11:34.000+01:00"
  >
    10 minutes ago
  </time>
</div>
`;

exports[`Time renders relative time 2`] = `
<div>
  <time
    class="whitespace-nowrap"
    datetime="2022-02-20T10:11:34.000+01:00"
    title="2022-02-20T10:11:34.000+01:00"
  >
    10 minutes ago
  </time>
</div>
`;

exports[`Time renders the time in different formats 1`] = `
<div>
  <time
    class="whitespace-nowrap"
    datetime="2022-02-20T10:11:34.000+01:00"
    title="2022-02-20T10:11:34.000+01:00"
  >
    2/20/2022, 10:11 AM
  </time>
</div>
`;

exports[`Time renders the time in different formats 2`] = `
<div>
  <time
    class="whitespace-nowrap"
    datetime="2022-02-20T10:11:34.000+01:00"
    title="2022-02-20T10:11:34.000+01:00"
  >
    2/20/2022
  </time>
</div>
`;
